### ISSUE TITLE
Trying to use pure C code on PC

### ISSUE THREAD

**Issue by jasonwu1977** [JeffyCN/ARM-KWS-demo#2]
Hi Jeffy, 
Thanks for your sharing, I have follow all your steps and want to make a simulate with pure C code with running int8 quantization DS-CNN KWS model, but i have met the several problem
1. I create a Accuracy check function in C, which I use around 3700 *.wav in google's test set
2. I Run with the code, with "ORIGINAL" DS-CNN *.weight, which is [https://github.com/JeffyCN/ML-KWS-for-MCU/blob/master/Deployment/Source/NN_C/DS_CNN/ds_cnn_weights.h](url), and i get very good result
# Use ORIGINAL weight (Validation Data set)
correct_cnt:3334, fail_cnt:369
Accuracy is 90.035103%
3. BUT, when i use my self-trained weight, act_max, and receive weight & bias & parameter sets
test with my code again, and got poor result 
# Use Self-Trained weight
correct_cnt:1753, fail_cnt:1950
Accuracy is 47.339993%

Do you have any idea with it? or are you interesting about talking with it?
by the way i am chinese also, can I ask you in wechat? Thanks a lot!!

**Comment by JeffyCN** [comment]
Sorry, i don't quite understand the issue here, could you give more details?
1/ Have you followed all the retaining steps in the README.md?
2/ Have you check the accuracy related logs in these steps?
Run "./train.sh" to train
Run "./test.sh" to check the accuracy
Run "./fold_batchnorm.sh" to fuse batch-norm layers
Run "./quant_dump.sh" to quantize and dump params

**Comment by jasonwu1977** [comment]
1. Yes, I follow all the re-train steps in the README.md
2. Yes I have the log

`root@8d2a69b04702:/srv/wu/dscnn/jeffy/ML-KWS-for-MCU# ./quant_test.sh
Using checkpoint work/DS_CNN/DS_CNN1/training/best/ds_cnn_9390.ckpt-16000_bnfused!
Running: python quant_test.py       --data_url=   --data_dir=/dataset/speech_command/   --wanted_words yes,no,up,down,left,right,on,off,stop,go   --clip_duration_ms 1000   --model_architecture ds_cnn   --testing_percentage 10   --validation_percentage 10   --training_percentage 80   --unknown_percentage 50      --model_size_info 5 64 10 4 2 2 64 3 3 1 1 64 3 3 1 1 64 3 3 1 1 64 3 3 1 1   --dct_coefficient_count 10   --window_size_ms 40   --window_stride_ms 20     --act_max 32 4 16 8 16 16 8 32 8 8 4 8 1  --checkpoint work/DS_CNN/DS_CNN1/training/best/ds_cnn_9390.ckpt-16000_bnfused
2020-03-06 08:10:51.157245: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2020-03-06 08:10:51.905434: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1432] Found device 0 with properties:
name: Quadro GV100 major: 7 minor: 0 memoryClockRate(GHz): 1.627
pciBusID: 0000:18:00.0
totalMemory: 31.75GiB freeMemory: 24.94GiB
2020-03-06 08:10:52.624550: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1432] Found device 1 with properties:
name: Quadro GV100 major: 7 minor: 0 memoryClockRate(GHz): 1.627
pciBusID: 0000:3b:00.0
totalMemory: 31.75GiB freeMemory: 20.07GiB
2020-03-06 08:10:52.625391: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1511] Adding visible gpu devices: 0, 1
2020-03-06 08:10:55.821183: I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-03-06 08:10:55.821346: I tensorflow/core/common_runtime/gpu/gpu_device.cc:988]      0 1
2020-03-06 08:10:55.821369: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 0:   N Y
2020-03-06 08:10:55.821380: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 1:   Y N
2020-03-06 08:10:55.822701: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 24178 MB memory) -> physical GPU (device: 0, name: Quadro GV100, pci bus id: 0000:18:00.0, compute capability: 7.0)
2020-03-06 08:10:55.823961: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:1 with 19439 MB memory) -> physical GPU (device: 1, name: Quadro GV100, pci bus id: 0000:3b:00.0, compute capability: 7.0)
2020-03-06 08:11:10.221513: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1511] Adding visible gpu devices: 0, 1
2020-03-06 08:11:10.221948: I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-03-06 08:11:10.221970: I tensorflow/core/common_runtime/gpu/gpu_device.cc:988]      0 1
2020-03-06 08:11:10.221992: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 0:   N Y
2020-03-06 08:11:10.222009: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 1:   Y N
2020-03-06 08:11:10.224253: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 24178 MB memory) -> physical GPU (device: 0, name: Quadro GV100, pci bus id: 0000:18:00.0, compute capability: 7.0)
2020-03-06 08:11:10.225133: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:1 with 19439 MB memory) -> physical GPU (device: 1, name: Quadro GV100, pci bus id: 0000:3b:00.0, compute capability: 7.0)
2020-03-06 08:11:39.184110: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1511] Adding visible gpu devices: 0, 1
2020-03-06 08:11:39.184390: I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-03-06 08:11:39.184411: I tensorflow/core/common_runtime/gpu/gpu_device.cc:988]      0 1
2020-03-06 08:11:39.184429: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 0:   N Y
2020-03-06 08:11:39.184442: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 1:   Y N
2020-03-06 08:11:39.185665: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 24178 MB memory) -> physical GPU (device: 0, name: Quadro GV100, pci bus id: 0000:18:00.0, compute capability: 7.0)
2020-03-06 08:11:39.186365: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:1 with 19439 MB memory) -> physical GPU (device: 1, name: Quadro GV100, pci bus id: 0000:3b:00.0, compute capability: 7.0)
INFO:tensorflow:Restoring parameters from work/DS_CNN/DS_CNN1/training/best/ds_cnn_9390.ckpt-16000_bnfused
DS-CNN_conv_1_weights_0 number of wts/bias: (10, 4, 1, 64) dec bits: 9 max: (0.15039062,0.14951202) min: (-0.08984375,-0.09078087)
DS-CNN_conv_1_biases_0 number of wts/bias: (64,) dec bits: 7 max: (0.71875,0.71982425) min: (-0.25,-0.25101247)
DS-CNN_conv_ds_1_dw_conv_depthwise_weights_0 number of wts/bias: (3, 3, 64, 1) dec bits: 5 max: (3.53125,3.5170197) min: (-3.1875,-3.1926253)
DS-CNN_conv_ds_1_dw_conv_biases_0 number of wts/bias: (64,) dec bits: 6 max: (1.78125,1.7788297) min: (-1.265625,-1.2623914)
DS-CNN_conv_ds_1_pw_conv_weights_0 number of wts/bias: (1, 1, 64, 64) dec bits: 7 max: (0.609375,0.6062388) min: (-0.609375,-0.6075925)
DS-CNN_conv_ds_1_pw_conv_biases_0 number of wts/bias: (64,) dec bits: 6 max: (1.53125,1.5255712) min: (-0.9375,-0.936763)
DS-CNN_conv_ds_2_dw_conv_depthwise_weights_0 number of wts/bias: (3, 3, 64, 1) dec bits: 5 max: (3.625,3.62093) min: (-1.5625,-1.5510896)
DS-CNN_conv_ds_2_dw_conv_biases_0 number of wts/bias: (64,) dec bits: 6 max: (1.515625,1.5079643) min: (-1.390625,-1.396535)
DS-CNN_conv_ds_2_pw_conv_weights_0 number of wts/bias: (1, 1, 64, 64) dec bits: 7 max: (0.6171875,0.6198071) min: (-0.703125,-0.7000993)
DS-CNN_conv_ds_2_pw_conv_biases_0 number of wts/bias: (64,) dec bits: 5 max: (2.03125,2.0169473) min: (-1.5625,-1.5561056)
DS-CNN_conv_ds_3_dw_conv_depthwise_weights_0 number of wts/bias: (3, 3, 64, 1) dec bits: 5 max: (2.875,2.8631408) min: (-2.4375,-2.4300873)
DS-CNN_conv_ds_3_dw_conv_biases_0 number of wts/bias: (64,) dec bits: 6 max: (1.828125,1.8263255) min: (-1.75,-1.7553374)
DS-CNN_conv_ds_3_pw_conv_weights_0 number of wts/bias: (1, 1, 64, 64) dec bits: 7 max: (0.75,0.7475313) min: (-0.703125,-0.7068947)
DS-CNN_conv_ds_3_pw_conv_biases_0 number of wts/bias: (64,) dec bits: 6 max: (1.09375,1.0916464) min: (-1.578125,-1.5749665)
DS-CNN_conv_ds_4_dw_conv_depthwise_weights_0 number of wts/bias: (3, 3, 64, 1) dec bits: 6 max: (1.34375,1.3496113) min: (-1.109375,-1.1073895)
DS-CNN_conv_ds_4_dw_conv_biases_0 number of wts/bias: (64,) dec bits: 6 max: (1.34375,1.3399452) min: (-1.5625,-1.5618994)
DS-CNN_conv_ds_4_pw_conv_weights_0 number of wts/bias: (1, 1, 64, 64) dec bits: 7 max: (0.8046875,0.8038199) min: (-0.7890625,-0.78548396)
DS-CNN_conv_ds_4_pw_conv_biases_0 number of wts/bias: (64,) dec bits: 6 max: (1.71875,1.7260395) min: (-0.578125,-0.58004713)
DS-CNN_fc1_weights_0 number of wts/bias: (64, 12) dec bits: 5 max: (1.1875,1.1757575) min: (-2.09375,-2.1067932)
DS-CNN_fc1_biases_0 number of wts/bias: (12,) dec bits: 9 max: (0.23046875,0.2304504) min: (-0.103515625,-0.1031305)
INFO:tensorflow:set_size=49231
INFO:tensorflow:Confusion Matrix:
 [[ 3049     0     0     0     0     0     0     0     0     0     0     0]
 [   12 14882    10    40    24    48    24    27    27    33    36    68]
 [    0    79  3207     6     0     0    26     0     0     0     0     0]
 [    0    88     4  2984     0    31     5     0     0     1     0    52]
 [    2    88     0     1  2707     0     0     0     9    76    17     3]
 [    7    86     1    30     3  2958     0     0     0     7     6    30]
 [    3   159     9     0     6     3  2866    17     0     4     2     0]
 [    4   124     0     2     0     0     5  2878     0     0     0     0]
 [    1   126     0     0    17     2     0     0  2915    41     0     6]
 [    1    68     0     0    39     0     0     0    29  2844     0     6]
 [    3   104     0     0    16    15     0     0     0     3  2891    10]
 [    7   127     1    42     1    16     1     0     1     8     8  3006]]
INFO:tensorflow:Training accuracy = 95.85% (N=49231)
INFO:tensorflow:set_size=5926
INFO:tensorflow:Confusion Matrix:
 [[ 371    0    0    0    0    0    0    0    0    0    0    0]
 [   2 1795    2    8    4    6    4    7    8    3    6    7]
 [   0   14  372    1    0    0    7    1    0    0    0    2]
 [   1   22    1  367    1    5    2    0    0    0    0    7]
 [   0   10    0    0  315    1    1    0    1   16    5    1]
 [   0   10    0   10    0  352    0    0    0    0    1    4]
 [   0   16    3    0    0    0  331    1    0    0    0    1]
 [   0   20    1    0    0    0    0  342    0    0    0    0]
 [   1   16    0    0    3    0    0    0  337    6    0    0]
 [   0   10    0    0   17    0    1    0    6  335    2    2]
 [   1   13    0    0   14    1    0    0    0    3  318    0]
 [   0   16    0    9    2    0    0    0    0    2    1  342]]
INFO:tensorflow:Validation accuracy = 94.11% (N=5926)
INFO:tensorflow:set_size=6519
INFO:tensorflow:Confusion Matrix:
 [[ 408    0    0    0    0    0    0    0    0    0    0    0]
 [   0 1948    1    7    4   13   11   12    9   12    3   17]
 [   0   17  397    3    0    0    1    0    0    0    1    0]
 [   0   15    0  375    0    5    1    0    0    1    0    8]
 [   0   16    0    0  389    1    0    0    3   13    2    1]
 [   0   21    1   13    0  362    1    0    1    0    0    7]
 [   1   18    4    0    1    0  387    1    0    0    0    0]
 [   0   19    0    0    0    0    5  371    0    0    1    0]
 [   0   19    0    0    1    4    0    0  356   14    0    2]
 [   1   15    0    0    9    0    0    0    4  368    0    5]
 [   0   15    0    0    7    2    0    0    0    1  385    1]
 [   0   17    0   14    0    4    0    0    0    0    0  367]]
INFO:tensorflow:Test accuracy = 93.77% (N=6519)
root@8d2a69b04702:/srv/wu/dscnn/jeffy/ML-KWS-for-MCU#`

**Comment by jasonwu1977** [comment]
As you can see, the result is very good, no problem, then i get the **weight & bias & parameters**
i think it generates from "act_max 32 4 16 8 16 16 8 32 8 8 4 8 1"
then i test on C code ( I write a test loop myself, which can test the accruacy with 3700 wav data)

but the result is bad, i doubt if anyone who test the result running with C-code ( or embedded) , or where i have made mistakes
anyway, thanks a lot for reply

**Comment by JeffyCN** [comment]
Okay, so maybe you can try to:
1/ Test the C-code with your training data set
2/ Try to run test.sh/quant_test.sh on your final validation data set(modify the related params in common.sh)
3/ Test the generated headers with the original ARM-KMS demo(C++) (might need some small adjustments)

**Comment by jasonwu1977** [comment]
Yes, for (2)
common.sh already modified
i just run quant_test.sh, and the result was good 
and i got generated header weight here

But for (3)
Test with with original ARM-KMS header weight, result is good, Accuracy is 90.035103%
test with my generated header weight, result is bad, Accruracy is 47.339993%

**Comment by jasonwu1977** [comment]
Jeffy, i think i figure out why
the first act_max **MUST** use 64 not 32

**Comment by JeffyCN** [comment]
Sorry, i think you missed my point..

So there are 4 models here:
a/ the trained one(generated from train.sh), only for tensorflow
b/ the fused one(fold_batchnorm.sh), only for tensorflow
c/ the tensorflow quantized one(quant_test.sh), only for tensorflow
d/ the finial quantized one(generated headers), only for CMSIS5

And there are 4 runners here:
e/ test scripts, using tensor flow, can run with a/ b/ c/
f/ original C++ demo, using ARM CMSIS5, can run with d/
g/ my C version demo, using ARM CMSIS5, can run with d/
h/ your custon C-version demo, using ARM CMSIS5, can run with d/

And there are 2 data sets:
i/ training data set, used as training data/validation data/test data by scripts
j/ the final validation data set, used by your custom test case(C-code)

If i'm reading this issue right, you said:
1/ using scripts: e/(tensorflow) + a/b/c(models) + i/(training data set) = good accuracy
2/ using h/(your C-code demo) + d/(quanted headers) + j/(final validation data set) = bad accuracy

My suggestions are:
1/ try h/(your C-code demo) + d/(quanted headers) + i/(training data set) 
2/ try e/(tensorflow scripts) + a/b/c(models) + j/(final validation data set) 
3/ try f/(original C++ demo) + d/(quanted headers) + j/(final validation data set)

**Comment by jasonwu1977** [comment]
Jeffy, I am using speech_commands_v0.02.tar.gz as dataset, which i just use testing_list.txt & validation_list.txt as dataset, both result on your suggestion 
1/ try h/(your C-code demo) + d/(quanted headers) + i/(training data set) was low
3/ try f/(original C++ demo) + d/(quanted headers) + j/(final validation data set) was high
(i can't try tensorflow with these dataset, the code looks complicated on pick up testing wav file)

I can see the result is, when i use ds_cnn, the act_max **"64** 0 0 0 0 0 0 0 0 0 0 0 0" the first one, must be 64, not 32, otherwise the accuracy will drop to half, i don't know why and i am doing more test with it, thanks for help and reply!

**Comment by JeffyCN** [comment]
Hmmm, the act_max array is about how much details to keep for each layer's activate data during quantizing to get a better accuracy, so actually it could be 32.

According to your quant_test.sh log, the tensorflow could get a decent result with that.

for test 2/, you can simply replace the dir setting in common.sh to change data set.

for the other tests, it looks like something wrong between original c++ version and the c version.
